cirq-google/cirq_google/ops/analog_detune_gates.py

```diff
@@ -0,0 +1,197 @@
+# Copyright 2025 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Define detuning gates for Analog Experiment usage."""
+from __future__ import annotations
+
+from typing import AbstractSet, Any, TYPE_CHECKING, TypeAlias
+
+import sympy
+import tunits as tu
+
+import cirq
+
+if TYPE_CHECKING:
+    import numpy as np
+
+# The gate is intended for the google internal use, hence the typing style
+# follows more on the t-unit + symbol instead of float + symbol style.
+ValueOrSymbol: TypeAlias = tu.Value | sympy.Basic
+FloatOrSymbol: TypeAlias = float | sympy.Basic
+
+
+@cirq.value_equality(approximate=True)
+class AnalogDetuneQubit(cirq.ops.Gate):
+    """A step function that steup a qubit to the freq according to analog model.
+    Pulse shape:
+    .. svgbob::
+      :align: center
+                   |   ,--------|---- amp (calculated from target freq using analog model)
+                   |  /         |
+        prev_amp---|-' - - - - -| - -
+        ^
+        |          |-w -|       |
+        |          |---length --|
+        |
+        --------------------------(calculated from previous qubit freq using analog model)
+    Because the step hold at amp with inifinite length. You need prev_freq to compensate
+    previous Detune gate. If not provide, no compensation will be applied, i.e. start from 0.
+    If the target_freq is None and prev_freq provided, this Detune gate will
+    reset the qubit freq back to absolute amp=0 according to prev_freq.
+    """
+
+    def __init__(
+        self,
+        length: ValueOrSymbol,
```

Comment:
Optional:  I wonder if it will be easier to just make it length_ns and float rather than fight against cirq's typing.  Sweeps etc are all defined as floats.

Reply:
@trondandersen

Reply:
I mean that the type is defined as floats in cirq.  It works with things other than floats some of the time.

Reply:
maybe I misunderstand the comment. Which line of code are you referring to? Here the type is `ValueOrSymbol`, which is `tu.Value | Symbol` and type checking is okay.

Reply:
What I mean is this https://quantumai.google/reference/python/cirq/TParamVal
cirq itself defines a parameter value as a float, which is used in sweeps all over:
https://github.com/quantumlib/Cirq/blob/v1.5.0/cirq-core/cirq/study/sweeps.py#L41

Not much of the code actually depends that this is a float, but some of it (like gate implementations) does, so we are always going to be fighting against the typing system if we are trying to support sweeps with units.

Reply:
I see. Yeah, i also feel there is always a fighting with value versus float convention. I think for this analog_experiment one, we will not have code about sweep to check-in. 

In cirq-google sweep, we have expanded the proto definition to support tunits. https://github.com/quantumlib/Cirq/blob/7fe741bfbce2a58f803ecb04d577bb45802fb2f4/cirq-google/cirq_google/api/v2/run_context.proto#L226-L250

So running the experiment with tunit is okay and we don't need to check-in code to fight with type system. But happy to find a middle ground if we do encounter type checking issues.